### PR TITLE
Update spec for cookie partition keys and partitioned storage keys

### LINKED ISSUE #75: Delete CHIPS cookies
When (deleting cookies)[https://privacycg.github.io/nav-tracking-mitigations/#clear-cookies-for-host], only cookies whose domain key matches are deleted. We should also delete cookies whose partition key is same-site-host with the host to be deleted.

### PR DESCRIPTION
**Pull request by svendlarsen**
Namely, when deleting cookies for a host:

- Preserve cookies for that host partitioned under other sites, but
- delete cookies partitioned under sites with that host.

And change non-cookie storage deletion:

- from looking at the storage key's origin
- to looking at the storage key's top-level site.

Fixes #75.

***
<a href="https://pr-preview.s3.amazonaws.com/svendlarsen/nav-tracking-mitigations/pull/78.html" title="Last updated on Jul 22, 2024, 6:21 PM UTC (7fd9126)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/privacycg/nav-tracking-mitigations/78/d6335ec...svendlarsen:7fd9126.html" title="Last updated on Jul 22, 2024, 6:21 PM UTC (7fd9126)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -755,8 +755,10 @@ spec.  It would be nice to unify these in the future.</p>
 To <dfn>clear cookies for host</dfn> given a [=host=] |host|, perform the
 following steps:
 
-1. Let |cookieList| be the set of cookies from the [=cookie store=] whose
-    domain attribute is a [=domain-match=] with |host|.
+1. Let |cookieList| be a set of cookies, initially empty.
+1. [=list/For each=] cookie |cookie| in the [=cookie store=], if either of the following is true, add |cookie| to |cookieList|:
+     - |cookie| is not partitioned, and |cookie|'s domain attribute is a [=domain-match=] with |host|; or
+     - |cookie| is partitioned, and |cookie|'s partition key is equal to |host|'s [=obtain a site|site=].
```
Comment (jyasskin) [inline index.bs:761]: This isn't worse than what was here before, so I don't object to merging it, but I'd really like to see "partitioned", "domain attribute", and "partition key" link to their definitions. I think "domain attribute" could link to https://datatracker.ietf.org/doc/html/rfc6265/#section-5.2.3 now. @johannhof / @dcthetall, is https://www.ietf.org/archive/id/draft-cutler-httpbis-partitioned-cookies-01.html the right citation for the partitioning aspects?

**Review by jyasskin — COMMENTED** [review]
This looks good enough from a spec point of view, but I'd also like @bvandersloot-mozilla to confirm that this fixes his bug report.

File: index.bs (hunk 1 above)
Comment (svendlarsen) [reply index.bs]: I've added dfn links for those three terms as you suggested, but happy to update if there's a better doc to link to for CHIPS terms.

**Review by bvandersloot-mozilla — APPROVED** [review]
:+1: This resolves the issue.

**Review by johannhof — APPROVED** [review]
So, I think overall, the way we'd rather do this in the future is to expose some interface for this in the cookies RFC (which defines the cookie store) as web platform specs don't necessarily have direct introspection into the cookie store. @annevk any thoughts?

With that said, I acknowledge that we're not fully there yet and so I'm okay with this as I think it's sketching out the correct steps. We can add this to the list for cookie layering work...

File: index.bs [hunk 2]
```diff
@@ -755,8 +760,10 @@ spec.  It would be nice to unify these in the future.</p>
 To <dfn>clear cookies for host</dfn> given a [=host=] |host|, perform the
 following steps:
 
-1. Let |cookieList| be the set of cookies from the [=cookie store=] whose
-    domain attribute is a [=domain-match=] with |host|.
+1. Let |cookieList| be a set of cookies, initially empty.
+1. [=list/For each=] cookie |cookie| in the [=cookie store=], if either of the following is true, add |cookie| to |cookieList|:
+     - |cookie| is not [=partitioned cookie|partitioned=], and |cookie|'s [=domain attribute=] is a [=domain-match=] with |host|; or
+     - |cookie| is [=partitioned cookie|partitioned=], and |cookie|'s [=partition key=] is equal to |host|'s [=obtain a site|site=].
```
[the remark this answers is not in the document]
Comment (annevk) [reply index.bs:766]: This doesn't work. "obtain a site" takes an origin, not a host.

And yeah, doing this would require a new hook in https://github.com/johannhof/draft-annevk-johannhof-httpbis-cookies/ and also some decision around whether to dispatch events I guess? Probably best to start thinking about that new and leave some XXX/TODO marker here and file an issue.

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: > Yeah, I think this might need to adapt to ACB.

What is ACB?

> This doesn't work. "obtain a site" takes an origin, not a host.

Thanks for the catch — I've updated this line.

> And yeah, doing this would require a new hook in https://github.com/johannhof/draft-annevk-johannhof-httpbis-cookies/ and also some decision around whether to dispatch events I guess? Probably best to start thinking about that n[o]w and leave some XXX/TODO marker here and file an issue.

Happy to add a note about the intersection with the linked draft, but it's not clear to me what exactly I should be noting.

[the remark this answers is not in the document]
Comment (annevk) [reply index.bs]: This still does not seem correct to me. Or is host always canonicalized to a registrable domain beforehand or some such?

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: All the hosts that this algorithm is called for are from the user activation map, which indeed only stores site hosts, but that connection is not obvious as written. I changed the algorithm name to "clear cookies for site host" for clarity.

[the remark this answers is not in the document]
Comment (annevk) [reply index.bs]: Hmm, I think it would be better to undo that commit (site host isn't really a thing) and add an assert that host's registrable domain is null.

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: > (site host isn't really a thing)

I see.

> add an assert that host's registrable domain is null.

Based on my reading of https://url.spec.whatwg.org/#example-host-psl, I think I should assert that `host`'s registrable domain is still equal to `host` (e.g., in the table, `example.com`'s registrable domain is `example.com`). But totally open to the possibility that I've misunderstood something.

[the remark this answers is not in the document]
Comment (annevk) [reply index.bs]: You're right, that's not going to work. But asserting it's itself is also not going to work as per the `gitHub.io` example. I guess we can assert that it's either equal to itself or null. That's still useful and would fail for inputs like `www.example.com`, which is mainly what we want to catch.

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: Good point, and thanks for going back and forth with me. Updated in latest push.

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: @annevk Have your concerns been addressed?

**Comment by wanderview** [comment]
I think this PR also needs to specify the deletion of partitioned non-cookie storage.  If non-cookie storage partitioning is not specified adequately yet we should at least add a note indicating that is the intent.

File: index.bs (hunk 2 above)
[the remark this answers is not in the document]
Comment (annevk) [reply index.bs]: I think so, but as Johann pointed out the ancestor-chain-bit (ACB) needs addressing and Ben is correct about storage in general.

[the remark this answers is not in the document]
Comment (wanderview) [reply index.bs]: Can we rephrase this to just look at the top-level site contained in the partition key?  That way ACB and any future additions to the partition key would still be handled correctly.

[the remark this answers is not in the document]
Comment (annevk) [reply index.bs]: I think that works.

And I think as long as this works similarly to the end user deleting an entire site there's no problem either. I got a bit worried about giving a top-level site the ability to delete cookies outside of its direct control, but I think it holds up.

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: > Can we rephrase this to just look at the top-level site contained in the partition key?

Added in latest push. Working on the local storage piece.

[the remark this answers is not in the document]
Comment (svendlarsen) [reply index.bs]: Updated the deletion of non-cookie storage to account for the addition of top-level site to storage keys in whatwg/storage#144.

**Review by wanderview — APPROVED** [review]

**Comment by svendlarsen** [comment]
It seems like we've reached agreement on this change, at least given the current state of related spec/standard changes. I'm going to merge; please note any new concerns by filing an issue. Thanks all for your input!

**Comment by wanderview** [comment]
I think you need to drop the phantom link to the top-level site for the partition key.  Getting this failure on trying to submit:

image
